security/fma-holocene.md

```diff
@@ -0,0 +1,196 @@
+
+This will lead to the consensus client hitting error conditions when it loads the frames from L1. According to the spec, frames will be dropped in order to maintain a contiguous frame queue, and out-of-order batches will be dropped at a later stage in the derivation pipeline. The safe chain would halt until the next batch can be resubmitted in order.
+
+#### Risk Assessment
+
+low severity / low likelihood
+
+#### Mitigations
+
+The batcher implementation could:
+
+- leverage nonce management to avoid sending transactions out of order in the first place (see [spec](https://specs.optimism.io/protocol/holocene/derivation.html?highlight=holocene#batcher-hardening))
+- be hardened so detect the chain halt and resubmit the frames which were dropped (see [spec](https://specs.optimism.io/protocol/holocene/derivation.html?highlight=holocene#batcher-hardening)) either:
+  - as a part of normal operation
+  - as a part of its startup behavior (i.e. to be triggered by a restart)
+  - as a part of an emergency mode triggered by an admin API, this could allow for manual intervention
+- continually check for contiguity in its internal state, and panic or reset if this is violated (possibly then triggering some recovery behavior)
+
+#### Detection
+
+L2 safe chain halt
+
+#### Recovery Path(s)
+
+We would need to get the batcher to resubmit the transaction sent out of order. No governance or hardfork needed to recover. A simple batcher restart would cause the batcher to re-batch the unsafe chain, which should lead to recovery. Moreover, we could also temporarily operate the batcher with reduced tx sending concurrency, which should avoid out of order txs.
+
+### Derivation Deadlock (either specification or op-node bug)
+
+#### Description
+
+It is possible that either i) the Holocene Derivation Specification or ii) the consensus client implementation has overlooked a corner case such that derivation simply halts when that corner case arises.
+
+#### Risk Assessment
+
+**high severity** / medium likelihood
+
+#### Mitigations
+
+Multi client testing should help surface any consensus client bugs, possibly including fuzzing of some description.
+We should backport as much insight as possible from the implementation (which is code) to the specs (which is just words). Here's an example of doing just that [holocene: fix frame queue specs of derivation](https://github.com/ethereum-optimism/specs/pull/386).
+
+#### Detection
+
+L2 safe chain halt
+
+#### Recovery Path(s)
+
+We would need to fix the the implementation via a hardfork. As a hotfix, we would probably also make an emergency release which migrates back to the old DP at the block where derivation is stuck. We might already implement something like this as a contingency, e.g. adding a holocene_deactivation_time that would then move back to old derivation if this time is set. We could then instruct node operators to set this flag to some value, providing a quick recovery path.
```

Comment:
Will we have that hotfix ready in advance? IIRC we had a contingency plan in place for ecotone. If there is a standard procedure, is it documented somewhere?

Reply:
This sounds mega painful with the fault proof system being live. We definitely don't want this to happen

Reply:
Action: link to relevant runbooks, ensure they are up to date and the plan is agreed to.

Reply:
IMO this belongs in the generic hardfork doc https://github.com/ethereum-optimism/design-docs/pull/131

Reply:
Move it here [68f0cd6](https://github.com/ethereum-optimism/design-docs/pull/127/commits/68f0cd6e7b1cdbf108feaf5044cd79b2a7e69798) but let's keep this thread unresolved until we have linked to a runbook (I think we may need to create a new runbook).

Reply:
Linked to the "chain" rewind runbook here [2f5dccc](https://github.com/ethereum-optimism/design-docs/pull/127/commits/2f5dcccf9868012e3c72261de80f2102177b2243)